modules/lang-painless/src/main/java/org/elasticsearch/painless/Locals.java

```diff
@@ -112,33 +92,20 @@ public static Locals newFunctionScope(Locals programScope, Type returnType, List
     }
 
     /** Creates a new main method scope */
-    public static Locals newMainMethodScope(Locals programScope, boolean usesScore, boolean usesCtx, int maxLoopCounter) {
-        Locals locals = new Locals(programScope, Definition.OBJECT_TYPE, MAIN_KEYWORDS);
-        // This reference.  Internal use only.
+    public static Locals newMainMethodScope(MainMethod mainMethod, Locals programScope, Iterable<DerivedArgument> usedDerivedArguments,
+            int maxLoopCounter) {
+        Locals locals = new Locals(programScope, Definition.OBJECT_TYPE, KEYWORDS);
+        // This reference. Internal use only.
         locals.defineVariable(null, Definition.getType("Object"), THIS, true);
 
-        // Input map of variables passed to the script.
-        locals.defineVariable(null, Definition.getType("Map"), PARAMS, true);
-
-        // Scorer parameter passed to the script.  Internal use only.
-        locals.defineVariable(null, Definition.DEF_TYPE, SCORER, true);
-
-        // Doc parameter passed to the script. TODO: Currently working as a Map, we can do better?
-        locals.defineVariable(null, Definition.getType("Map"), DOC, true);
-
-        // Aggregation _value parameter passed to the script.
-        locals.defineVariable(null, Definition.DEF_TYPE, VALUE, true);
-
-        // Shortcut variables.
-
-        // Document's score as a read-only double.
-        if (usesScore) {
```

Comment:
And these are now derived from the array of `DerivedArguments` passed to the `Compiler`.